state/protocol/protocol_state/state/mutable_protocol_state_test.go

```diff
@@ -0,0 +1,650 @@
+//
+// This test also verifies that the `MutableProtocolState` does not engage in step (iii) before the completing step (ii) on all state machines.
+func (s *StateMutatorSuite) Test_StateMachineProcessingServiceEventsFails() {
+	workingStateMachine := *protocol_statemock.NewOrthogonalStoreStateMachine[protocol.KVStoreReader](s.T())
+	workingStateMachine.On("EvolveState", mock.MatchedBy(emptySlice[flow.ServiceEvent]())).Return(nil).Once()
+
+	exception := errors.New("exception")
+	failingStateMachine := *protocol_statemock.NewOrthogonalStoreStateMachine[protocol.KVStoreReader](s.T())
+	failingStateMachine.On("EvolveState", mock.MatchedBy(emptySlice[flow.ServiceEvent]())).Return(exception).Once()
+
+	s.Run("failing state machine is last", func() {
+		s.kvStateMachines[0], s.kvStateMachines[1] = workingStateMachine, failingStateMachine //nolint:govet
+		_, dbUpdates, err := s.mutableState.EvolveState(s.candidate.ParentID, s.candidate.View, []*flow.Seal{})
+		require.Error(s.T(), err)
+		require.False(s.T(), protocol.IsInvalidServiceEventError(err))
+		require.True(s.T(), dbUpdates.IsEmpty())
+	})
+
+	failingStateMachine.On("EvolveState", mock.MatchedBy(emptySlice[flow.ServiceEvent]())).Return(exception).Once()
+	s.Run("failing state machine is first", func() {
+		s.kvStateMachines[0], s.kvStateMachines[1] = failingStateMachine, workingStateMachine //nolint:govet
+		_, dbUpdates, err := s.mutableState.EvolveState(s.candidate.ParentID, s.candidate.View, []*flow.Seal{})
+		require.Error(s.T(), err)
+		require.False(s.T(), protocol.IsInvalidServiceEventError(err))
+		require.True(s.T(), dbUpdates.IsEmpty())
+	})
+}
+
+// Test_StateMachineBuildFails verifies that errors are escalated to the caller that any of the sub-state machines return
+// during their `Build` step. Specifically, the state machine should handle invalid events internally and only escalate internal
+// errors in case of an irrecoverable problem.
+func (s *StateMutatorSuite) Test_StateMachineBuildFails() {
+	workingStateMachine := *protocol_statemock.NewOrthogonalStoreStateMachine[protocol.KVStoreReader](s.T())
+	workingStateMachine.On("EvolveState", mock.MatchedBy(emptySlice[flow.ServiceEvent]())).Return(nil).Twice()
+	workingStateMachine.On("Build").Return(transaction.NewDeferredBlockPersist(), nil).Maybe()
+
+	exception := errors.New("exception")
+	failingStateMachine := *protocol_statemock.NewOrthogonalStoreStateMachine[protocol.KVStoreReader](s.T())
+	failingStateMachine.On("EvolveState", mock.MatchedBy(emptySlice[flow.ServiceEvent]())).Return(nil).Twice()
+	failingStateMachine.On("Build").Return(nil, exception).Once()
+
+	s.Run("failing state machine is last", func() {
+		s.kvStateMachines[0], s.kvStateMachines[1] = workingStateMachine, failingStateMachine //nolint:govet
+		_, dbUpdates, err := s.mutableState.EvolveState(s.candidate.ParentID, s.candidate.View, []*flow.Seal{})
+		require.Error(s.T(), err)
```

Comment:
```suggestion
		require.ErrorIs(s.T(), err, exception)
```

Reply:
done in https://github.com/onflow/flow-go/pull/5644/commits/35454a2d84ea0df3411f41aa94cea23110400d9f